### PR TITLE
Authentication Part 1

### PR DESCRIPTION
**Pull request by impguard**
Attempt: 9001

First authentication implementation!

You can now login with a user and complete the registration process. You are also properly redirected to login (or not) depending on if you've logged in recently!

Authentication information is NOT currently checked anywhere else or in the backend. Also no global user information is stored. This is coming in part 2. Wanted to get the cucumber problems out of the way first with regards to properly navigating to the right pages despite our website kicking you around.

### CODE REVIEW

File: app/assets/javascripts/startpage/router.coffee [hunk 1]
```diff
@@ -8,6 +8,16 @@ router.config(["$stateProvider", "$urlRouterProvider", ($stateProvider, $urlRout
         url: "/"
         templateUrl: "/assets/login.html"
         controller: "LoginController"
+        resolve: {
+            loggedIn: ["Authentication", (Authentication) ->
+                return Authentication.isLoggedIn(false).then(
+                    (result) ->
+                        return true
```
Comment (whsieh) [inline app/assets/javascripts/startpage/router.coffee]: maybe `(result) -> true`?

File: spec/javascripts/controllers/surveyTopicsSpec.coffee [hunk 2]
```diff
@@ -7,7 +7,6 @@ describe "SurveyTopicsController", ->
             icon: "/fake/path",
             id: 1
         }]);
-        @http.expectGET("/api/v1/topics")
```
Comment (whsieh) [inline spec/javascripts/controllers/surveyTopicsSpec.coffee:10]: I think this should still be here

**Comment by impguard** [comment]
Fixed! Plus one it now peoples!

**Comment by nickfong** [comment]
Functionality is good, didn't throughly look at code but I didn't see anything obviously weird.  +1

File: app/assets/javascripts/startpage/surveySummary.coffee [hunk 3]
```diff
@@ -23,6 +26,21 @@ surveySummary.controller("SurveySummaryController", ["$scope", "$http", "$state"
...
+                    $cookieStore.put("user", user)
+                    $window.location.href = "/#/"
+                else
+                    $scope.status = if data then data["error"] else "Oops, an error occurred."
+            (reason) ->
+                $scope.status = "Oops, an error occurred. Try again!"
+        ).finally(() ->
+            $scope.submitRegistrationText = "Try Again"
+        )
+
```
Comment (leohtkam) [inline app/assets/javascripts/startpage/surveySummary.coffee:47]: Should disable the submission button during the registration process.

File: app/assets/javascripts/shared/api.coffee [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    this.requestProfileByUID = (userId) ->
+        request = generateRequest("user/#{userId}/profile")
+        $http.get(request)
+
+    this.updateProfileByUID = (userId, username, avatar, blurb, hero, spectrum) ->
+        request = generateRequest("user/#{userId}/profile")
+        params = { username: username, avatar: avatar, political_blurb: blurb, political_hero: hero, political_spectrum: spectrum}
+        $http.post(request, params)
+
+    this.requestArenasByUser = (userId) ->
```
Comment (leohtkam) [inline app/assets/javascripts/shared/api.coffee]: Probably want to change to requestConversationByUser instead.

**Comment by leohtkam** [comment]
+1 when you fix the thing I commented on.

**Pull request merged**